src/psyclone/psyir/transformations/hoist_local_arrays_trans.py

```diff
@@ -173,22 +177,60 @@ def apply(self, node, options=None):
             # new memory allocation statement.
             dim_list = [Range.create(dim.lower.copy(), dim.upper.copy())
                         for dim in orig_shape]
-            arefs.append(ArrayReference.create(sym, dim_list))
-
-        freader = FortranReader()
-        fwriter = FortranWriter()
-        # TODO #1366: we have to use a CodeBlock in order to query whether or
-        # not the array has been allocated already.
-        code = f"allocated({automatic_arrays[0].name})"
-        expr = freader.psyir_from_expression(code, node.symbol_table)
-        if_expr = UnaryOperation.create(UnaryOperation.Operator.NOT, expr)
-        # TODO #1366: we also have to use a CodeBlock for the allocate().
-        alloc_arg = ",".join(fwriter(aref) for aref in arefs)
-        body = [freader.psyir_from_statement(f"allocate({alloc_arg})",
-                                             node.symbol_table)]
-        # Insert the conditional allocation at the start of the supplied
-        # routine.
-        node.children.insert(0, IfBlock.create(if_expr, body))
+            aref = ArrayReference.create(sym, dim_list)
+
+            # TODO #1366: we have to use a CodeBlock in order to query whether
+            # or not the array has been allocated already.
+            code = f"allocated({sym.name})"
+            allocated_expr = freader.psyir_from_expression(code, node.symbol_table)
+            if_expr = UnaryOperation.create(
+                        UnaryOperation.Operator.NOT, allocated_expr)
+
+            # Add runtime checks to verify that the boundaries haven't changed
+            # (we skip literals as we know they can't have changed)
+            check_added = False
+            for idx, dim in enumerate(orig_shape):
+                if not isinstance(dim.lower, Literal):
+                    expr = BinaryOperation.create(
+                            BinaryOperation.Operator.NE,
+                            BinaryOperation.create(
+                                BinaryOperation.Operator.LBOUND,
+                                Reference(sym),
+                                Literal(str(idx+1), INTEGER_TYPE)),
+                            dim.lower.copy())
+                    if_expr = BinaryOperation.create(
```

Comment:
This confused me for a while so please could you add a comment saying something like "create() adds the new comparison as a child of the existing 'if not allocated' expression" and again at L213 below.

Reply:
I added the comments and also renamed if_expr to cond_expr as it really is just the conditional statement inside the if.
(I also fixed the doctest for `python -m doctest src/psyclone/psyir/transformations/hoist_local_arrays_trans.py` to pass)